File: src/system/announcements.ts
/**
 * src/system/announcements.ts — Announcement & Bulletin Engine
 */

import fs from 'node:fs';
import path from 'node:path';
import type { AnnouncementItem, AnnouncementState, AnnouncementCategory } from './types.ts';

export class AnnouncementEngine {
  private projectDir: string;
  private announcementsFile: string;
  private stateFile: string;

  constructor(projectDir: string = '.') {
    this.projectDir = path.resolve(projectDir);
    this.announcementsFile = path.join(this.projectDir, '.announcements.json');
    this.stateFile = path.join(this.projectDir, '.announcements_seen.json');
  }

  private getDefaultAnnouncements(): AnnouncementItem[] {
    return [
      {
        id: 'ann_v110_engines',
        title: 'Universal System Engines Suite Online',
        body: 'AgenticWorkflow v1.1.0 now integrates 9 core toolchain engines: Auto-Updater, Auto-Installer, Refresher, Doctor, Health Engine, Dependencies Engine, Notifications, Announcements, and Version Tracker with dual-runtime parity.',
        category: 'FEATURE',
        date: '2026-09-06',
        version: '1.1.0',
        priority: 'high'
      },
      {
        id: 'ann_toon_v41',
        title: 'TOON Protocol v4.1 Released',
        body: 'Token-Oriented Object Notation (v4.1) delivers 30-60% token compression across all multi-agent dialogues and workflow telemetry.',
        category: 'UPDATE',
        date: '2026-09-01',
        version: '1.0.5',
        priority: 'normal'
      },
      {
        id: 'ann_uahf_online',
        title: 'Universal Agentic Hooks Framework (UAHF)',
        body: 'UAHF governs Claude Code, Cursor, Codex, OpenCode, and shell processes with zero-drift safety policies.',
        category: 'FEATURE',
        date: '2026-08-25',
        version: '1.0.0',
        priority: 'normal'
      }
    ];
  }

  private getState(): AnnouncementState {
    try {
      if (fs.existsSync(this.stateFile)) {
        const raw = JSON.parse(fs.readFileSync(this.stateFile, 'utf-8'));
        const seenIds = Array.isArray(raw.seenIds) ? raw.seenIds : (Array.isArray(raw.seen_ids) ? raw.seen_ids : []);
        return { seenIds, lastChecked: raw.lastChecked || raw.last_checked || new Date().toISOString() };
      }
    } catch {
      // Ignore
    }
    return { seenIds: [], lastChecked: new Date().toISOString() };
  }

  private saveState(state: AnnouncementState): void {
    try {
      const data = {
        seenIds: state.seenIds,
        seen_ids: state.seenIds,
        lastChecked: state.lastChecked
      };
      fs.writeFileSync(this.stateFile, JSON.stringify(data, null, 2));
    } catch {
      // Ignore
    }
  }

  public listAll(): Array<AnnouncementItem & { seen: boolean }> {
    let list = this.getDefaultAnnouncements();
    if (fs.existsSync(this.announcementsFile)) {
      try {
        const custom = JSON.parse(fs.readFileSync(this.announcementsFile, 'utf-8'));
        if (Array.isArray(custom)) {
          list = [...custom, ...list];
        }
      } catch {
        // Ignore
      }
    }

    const state = this.getState();
    return list.map(item => ({
      ...item,
      seen: state.seenIds.includes(item.id)
    }));
  }

  public getUnread(): AnnouncementItem[] {
    return this.listAll().filter(a => !a.seen);
  }

  public markAsRead(id: string): void {
    const state = this.getState();
    if (!state.seenIds.includes(id)) {
      state.seenIds.push(id);
      this.saveState(state);
    }
  }

  public markAllAsRead(): void {
    const all = this.listAll();
    const state: AnnouncementState = {
      seenIds: all.map(a => a.id),
      lastChecked: new Date().toISOString()
    };
    this.saveState(state);
  }

  public addAnnouncement(item: AnnouncementItem): void {
    let custom: AnnouncementItem[] = [];
    if (fs.existsSync(this.announcementsFile)) {
      try {
        custom = JSON.parse(fs.readFileSync(this.announcementsFile, 'utf-8'));
      } catch {
        // Ignore
      }
    }
    custom.unshift(item);
    fs.writeFileSync(this.announcementsFile, JSON.stringify(custom, null, 2));
  }

  public renderBroadcastBanner(): string | null {
    const unread = this.getUnread();
    if (unread.length === 0) return null;

    const top = unread[0];
    const color = '\x1b[35m'; // Magenta
    const reset = '\x1b[0m';
    const bold = '\x1b[1m';

    return `\n${color}📢 [Announcement]${reset} ${bold}${top.title}${reset} (${top.date})\n   ${top.body.substring(0, 100)}${top.body.length > 100 ? '...' : ''}\n   Run 'agentic-workflow announcements' to view all.\n`;
  }
}

File: core/system/announcements.py
"""
core/system/announcements.py — Python Announcement & Bulletin Engine
"""

import json
from pathlib import Path
from typing import Dict, Any, List


class AnnouncementEngine:
    def __init__(self, project_dir: str = "."):
        self.project_dir = Path(project_dir).resolve()
        self.state_file = self.project_dir / ".announcements_seen.json"

    def get_default_announcements(self) -> List[Dict[str, Any]]:
        return [
            {
                "id": "ann_v110_engines",
                "title": "Universal System Engines Suite Online",
                "body": "AgenticWorkflow now features 9 core toolchain engines with dual-runtime parity.",
                "category": "FEATURE",
                "date": "2026-09-06",
                "version": "1.1.0",
                "priority": "high"
            },
            {
                "id": "ann_toon_v41",
                "title": "TOON Protocol v4.1 Released",
                "body": "Token-Oriented Object Notation (v4.1) delivers 30-60% token compression.",
                "category": "UPDATE",
                "date": "2026-09-01",
                "version": "1.0.5",
                "priority": "normal"
            }
        ]

    def get_seen_ids(self) -> List[str]:
        if self.state_file.exists():
            try:
                with open(self.state_file, "r", encoding="utf-8") as f:
                    return json.load(f).get("seen_ids", [])
            except Exception:
                return []
        return []

    def list_all(self) -> List[Dict[str, Any]]:
        seen = set(self.get_seen_ids())
        anns = self.get_default_announcements()
        for a in anns:
            a["seen"] = a["id"] in seen
        return anns

    def get_unread(self) -> List[Dict[str, Any]]:
        return [a for a in self.list_all() if not a.get("seen")]

    def mark_as_read(self, ann_id: str):
        seen = self.get_seen_ids()
        if ann_id not in seen:
            seen.append(ann_id)
            try:
                with open(self.state_file, "w", encoding="utf-8") as f:
                    json.dump({"seen_ids": seen}, f, indent=2)
            except Exception:
                _saved = False

    def mark_all_as_read(self):
        all_ids = [a["id"] for a in self.list_all()]
        try:
            with open(self.state_file, "w", encoding="utf-8") as f:
                json.dump({"seen_ids": all_ids}, f, indent=2)
        except Exception:
            _saved = False
